server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/rfc/model/Capability.scala

```diff
@@ -0,0 +1,41 @@
+/** *************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                 *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package scala.org.apache.james.jmap.rfc.model
+
+import java.net.URI
+
+final case class CapabilityIdentifier(value: URI)
+
+sealed trait Capability
+
+final case class CoreCapability(maxSizeUpload: Int, maxConcurrentUpload: Int,
```

Comment:
what `extends AnyVal` mean?

Reply:
It mean that it is a value class. So the value will not be boxed. For the compiler it will be statically checked, but at the JVM level it will be the primitive type of the inner value without overhead.

Reply:
> case class Size(value: Int) extends AnyVal

I can create a type for Size but:
1. It cannot extend AnyVal, because my Size's constructor has checks for unsigned int and it breaks the requirements to be extending AnyVal
2. Do I need to do it for other types? Beside of size, I also have depth, number all represented in Int ?

Reply:
> Do I need to do it for other types?

Strong typing everywhere. It's cheap with case classes.

Reply:
These ints have restinctions:

```
The Int and UnsignedInt Data Types

Where Int is given as a data type, it means an integer in the range -2^53+1 <= value <= 2^53-1, the safe range for integers stored in a floating-point double, represented as a JSON Number.

Where UnsignedInt is given as a data type, it means an Int where the value MUST be in the range 0 <= value <= 2^53-1.
```

We do as far as I remember rely on a `Number` class in jmap-draft...

Reply:
We may use an implicit to convert int/long in `UnsignedInt` everywhere to avoid verbosity

Reply:
> It cannot extend AnyVal, because my Size's constructor has checks for unsigned int and it breaks the requirements to be extending AnyVal

@remk correct me if I'm wrong, but you can put the checks in the companion object `apply` method instead of in initialization code, right?